s3/s3_operations.cpp

```diff
@@ -1210,7 +1217,22 @@ namespace irods_s3 {
             return SUCCESS();
         }
 
+        // issue 1978 - If they called create without close, the object will not yet exist in s3.
+        // To be consistent with POSIX, return size of 0.
```

Comment:
Would we need to do any of this if `replica_close` closed the L3 descriptor before finalizing the replica?

In `rsDataObjClose`, we call `close_physical_file_and_throw_on_failure` before stat'ing the replica or doing anything with the catalog. In `replica_close`, we update the size and status of the replica in the catalog, and *then* close the physical file.

If the physical file was closed first, we wouldn't need to do this because the close operation would be called before we perform the stat and supposedly the object would exist and we could return the true size (which would be 0), right?

Reply:
For reference: https://github.com/irods/irods/blob/ed98d6d66e861f6ace4f72769e8f251856153f4f/plugins/api/src/replica_close.cpp#L431-L440

The suggestion is that perhaps we could move that block up above the block where the catalog updates occur: https://github.com/irods/irods/blob/ed98d6d66e861f6ace4f72769e8f251856153f4f/plugins/api/src/replica_close.cpp#L370-L371

Also for reference, here is the function in `rsDataObjClose`: https://github.com/irods/irods/blob/ed98d6d66e861f6ace4f72769e8f251856153f4f/server/api/src/rsDataObjClose.cpp#L586-L639 The very first thing that happens is that the physical file is closed.

Reply:
To build on top of this, we could provide an additional option on close that instructs it to swizzle the closing of the file on disk. Then the S3 resource can leverage that since it's special. Everything else can continue to operate as originally defined.

Reply:
Would this mean holding off in 1978 until 4.3.1 and 4.2.12?

Reply:
In that case, I will rollback the change for 1978 for now.